### PR TITLE
Add base implementation of native openID connect

### PR DESCRIPTION
**Pull request by DanMossa**
This pretty much overtakes this PR in draft
https://github.com/supabase-community/gotrue-dart/pull/57/files

I basically rewrote https://github.com/supabase/gotrue-js/pull/207/files in dart.

Once this is merged in I need to update [Supabase](https://github.com/supabase-community/supabase-dart)
Which will then require me to update https://github.com/supabase-community/supabase-flutter

### CODE REVIEW

File: lib/src/opent_id_connect_credentials.dart [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+///  The nonce is a random value generated by the developer (= yourself)
+///  before the initial grant is started.
+///  You should check the OpenID Connect specification for details.
+///  https://openid.net/developers/specs/
+///
+/// `provider` is the provider of the user.
+///
+/// `clientID` is the clientID of the user.
+///
+/// `issuer` is the issuer of the user.
```
Comment (bdlukaa) [inline lib/src/opent_id_connect_credentials.dart:14]: The documentation should be at each field

File: lib/src/opent_id_connect_credentials.dart [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+class OpenIDConnectCredentials {
+  final String idToken;
+  final String nonce;
+  final Provider? provider;
+  final String? clientId;
+  final String? issuer;
+
+  OpenIDConnectCredentials({
```
Comment (bdlukaa) [inline lib/src/opent_id_connect_credentials.dart:22]:
```suggestion
  cosnt OpenIDConnectCredentials({
```

File: lib/src/gotrue_client.dart [hunk 3]
```diff
@@ -99,32 +100,35 @@ class GoTrueClient {
...
+    if (email != null && password != null) {
+      return _handleEmailSignIn(email, password, options: options);
+    }
+    if (phone != null && password == null) {
+      final response = await api.sendMobileOTP(phone);
+      return GotrueSessionResponse(error: response.error);
+    }
+    if (phone != null && password != null) {
+      return _handlePhoneSignIn(phone, password);
+    }
```
Comment (bdlukaa) [inline lib/src/gotrue_client.dart:121]: This should be how it was before

**Review by bdlukaa — CHANGES_REQUESTED** [review]
Please format files using `dart format .`

File: lib/src/opent_id_connect_credentials.dart (hunk 1 above)
Comment (DanMossa) [reply lib/src/opent_id_connect_credentials.dart:14]: Like right each field?

File: lib/src/gotrue_client.dart (hunk 3 above)
Comment (DanMossa) [reply lib/src/gotrue_client.dart:121]: I kind of disagree. The way it previously was, was full of nested if statements that were hard to follow. This is also how the gotrue-js does it.

**Review by bdlukaa — APPROVED** [review]

**Comment by DanMossa** [comment]
Great! Ready to merge when you're ready @bdlukaa .
This fixes https://github.com/supabase-community/gotrue-dart/issues/27
and let's me get started on 
https://github.com/supabase-community/supabase-flutter/issues/5

**Comment by DanMossa** [comment]
@bdlukaa What's the process of bumping this to a new version so it can be used in supabase-dart?

**Comment by bdlukaa** [comment]
@dshukertjr usually do the releases of all dart packages. for now, you can use git dependencies

**Comment by hmarat** [comment]
@bdlukaa I do in this way, but I see  null logs anywhere(except my idToken got with google_sign_in) 
image

I don't know what's wrong here. Configs also are done correctly.
